narwhals/dataframe.py

```diff
@@ -64,32 +65,16 @@ def _from_compliant_dataframe(self, df: Any) -> Self:
             level=self._level,
         )
 
+    @abstractmethod
+    def _extract_compliant(self, arg: Any) -> Any:
+        raise NotImplementedError
+
     def _flatten_and_extract(self, *args: Any, **kwargs: Any) -> Any:
         """Process `args` and `kwargs`, extracting underlying objects as we go."""
         args = [self._extract_compliant(v) for v in flatten(args)]  # type: ignore[assignment]
         kwargs = {k: self._extract_compliant(v) for k, v in kwargs.items()}
         return args, kwargs
 
-    def _extract_compliant(self, arg: Any) -> Any:
```

Comment:
instead of defining this in BaseFrame, we define it in DataFrame and LazyFrame so they can have different behaviours